testing-common/src/main/groovy/io/opentelemetry/instrumentation/test/RetryOnAddressAlreadyInUseTrait.groovy

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.test
+
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+/**
+ * A trait for retrying operation when it fails with "java.net.BindException: Address already in use"
+ */
+trait RetryOnAddressAlreadyInUseTrait {
+  private static final Logger log = LoggerFactory.getLogger(RetryOnAddressAlreadyInUseTrait)
+
+  /**
+   * This is used by setupSpec() methods to auto-retry setup that depends on finding and then using
+   * an available free port, because that kind of setup can fail sporadically if the available port
+   * gets re-used between when we find the available port and when we use it.
+   *
+   * @param closure the groovy closure to run with retry
+   */
+  static void withRetryOnAddressAlreadyInUse(Closure<?> closure) {
+    withRetryOnAddressAlreadyInUse(closure, 3)
+  }
+
+  static void withRetryOnAddressAlreadyInUse(Closure<?> closure, int numRetries) {
```

Comment:
We already have three retry mechanisms available, gradle-retry-plugin, spock `@Retry`, and the GitHub actions workflow retry, would be great if we can avoid yet another retry mechanism.

Is it possible to use lazy-loaded singletons instead to take advantage of gradle-retry-plugin?

https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1751

Reply:
This is not new code, it was extracted from AgentTestRunner as @mateuszrzeszutek suggested. Anyway ideally such retry should not be needed. To me it seems that for this retry to happen the port returned from `PortUtils.randomOpenPort` is either not usable (for example `PortUtils.randomOpenPort` uses `socket.setReuseAddress(true);` but if server does not then I guess server could fail to bind to that port) or someone else managed to pick the exact same port. It would be best to have each test worker to have its own separate port range that excludes ephemeral ports. Something like https://github.com/gradle/gradle/blob/master/subprojects/core/src/testFixtures/groovy/org/gradle/util/ports/FixedAvailablePortAllocator.groovy As this is not a new issue I would prefer for it to be worked on in a separate pr.